src/FrameBuffer.cpp

```diff
@@ -1281,18 +1282,126 @@ void _writeToRdram(TSrc* _src, TDst* _dst, TDst(*converter)(TSrc _c), TSrc _test
 	for (; y < _height; ++y) {
 		for (u32 x = 0; x < _width && numStored < _numPixels; ++x) {
 			c = _src[x + (_height - y - 1)*_width];
 			if (c != _testValue)
-				_dst[(x + dsty*_width) ^ _xor] = converter(c);
+				_dst[(x + y*_width) ^ _xor] = converter(c);
 			++numStored;
 		}
-		++dsty;
+	}
+}
+
+template <typename TSrc>
+void _writeToRdram16(TSrc* _src, u16* _dst, u16(*converter)(TSrc _c), TSrc _testValue, u32 _width, u32 _height, u32 _numPixels, u32 _startAddress, u32 _bufferAddress, u32 _bufferSize)
+{
+	u32 chunkStart = ((_startAddress - _bufferAddress) >> (_bufferSize - 1)) % _width;
+	if (chunkStart % 2 != 0) {
+		--chunkStart;
+		--_dst;
+		++_numPixels;
+	}
+
+	u32 numStored = 0;
+	u32 y = 0;
+	TSrc c1, c2;
+	u16 val1, val2;
+
+	u32* _dst32 = reinterpret_cast<u32*>(_dst);
+    u32 dst32index = 0;
+    u32 heightOffset;
+
+	if (chunkStart > 0) {
+		for (u32 x = chunkStart; x < _width; x += 2) {
+		    heightOffset = _height - y - 1;
+			c1 = _src[x + heightOffset*_width];
+		    c2 = _src[x + 1 + heightOffset*_width];
+		    val1 = c1 != _testValue ? converter(c1) : _dst[numStored + 1];
+		    val2 = c2 != _testValue ? converter(c2) : _dst[numStored];
+
+            _dst32[dst32index] = (val1 << 16) | val2;
+
+			numStored += 2;
+			++dst32index;
+		}
+		++y;
+	}
+
+	for (; y < _height; ++y) {
+		for (u32 x = 0; x < _width && numStored < _numPixels; x += 2) {
+		    heightOffset = _height - y - 1;
+			c1 = _src[x + heightOffset*_width];
+		    c2 = _src[x + 1 + heightOffset*_width];
+		    val1 = c1 != _testValue ? converter(c1) : _dst[numStored + 1];
+		    val2 = c2 != _testValue ? converter(c2) : _dst[numStored];
+
+            _dst32[dst32index] = (val1 << 16) | val2;
+
+			numStored += 2;
+			++dst32index;
+		}
+	}
+}
+
+template <typename TSrc>
+void _writeToRdram8(TSrc* _src, u8* _dst, u8(*converter)(TSrc _c), TSrc _testValue, u32 _width, u32 _height, u32 _numPixels, u32 _startAddress, u32 _bufferAddress, u32 _bufferSize)
+{
+	u32 chunkStart = ((_startAddress - _bufferAddress) >> (_bufferSize - 1)) % _width;
+	if (chunkStart % 2 != 0) {
+		--chunkStart;
+		--_dst;
+		++_numPixels;
+	}
+
+	u32 numStored = 0;
+	u32 y = 0;
+
+	TSrc c1, c2, c3, c4;
+	u16 val1, val2, val3, val4;
+
+	u32* _dst32 = reinterpret_cast<u32*>(_dst);
+    u32 dst32index = 0;
+    u32 heightOffset;
+
+	if (chunkStart > 0) {
+		for (u32 x = chunkStart; x < _width; x += 4) {
+		    heightOffset = _height - y - 1;
+			c1 = _src[x + heightOffset*_width];
+		    c2 = _src[x + 1 + heightOffset*_width];
+			c3 = _src[x + 2 + heightOffset*_width];
+		    c4 = _src[x + 3 + heightOffset*_width];
+		    val1 = c1 != _testValue ? converter(c1) : _dst[numStored + 3];
+		    val2 = c2 != _testValue ? converter(c2) : _dst[numStored + 2];
+		    val3 = c3 != _testValue ? converter(c3) : _dst[numStored + 1];
+		    val4 = c4 != _testValue ? converter(c4) : _dst[numStored];
+
+            _dst32[dst32index] = (val1 << 24) | (val2 << 16) | (val3 << 8) | val4;
```

Comment:
I really hope I don't have this reversed, so I need find a game that triggers this to make sure it's right.